File: web/static/js/reducers/projects.js
const initialState = {
    projects: null,
    project: null,
    totalPages: null,
    currentPage: 1,
    search: null,
    lastUpdated: null,
    errors: null
};

export default function reducer(state = initialState, action = {}) {
    switch (action.type) {
    case 'LOAD_PROJECTS':
        return {
            ...state,
            project: null,
            projects: action.data.projects,
            search: action.data.search,
            currentPage: action.data.currentPage,
            totalPages: action.data.totalPages,
            lastUpdated: action.data.lastUpdated,
            errors: null
        };
    case 'LOAD_PROJECT':
        return {
            ...state,
            project: action.data.project,
            errors: null
        };
    case 'UPDATE_PROJECT':
        return {
            ...state,
            project: action.data.project,
            errors: action.data.errors
        };
    case 'SET_SEARCH':
        return {
            ...state,
            search: action.search
        };
    default:
        return state;
    }
}

File: lib/gt/manager/projects.ex
defmodule Gt.Manager.Projects do
    alias Gt.Model.Project
    alias Gt.Manager.Date, as: GtDate

    @page_size 10

    def load_projects(page, search \\ nil) do
        projects = Project
        projects = case is_nil(search) do
            true -> projects
            false -> projects |> Project.filter_by_title(search)
        end
        %{
            projects: projects |> Project.page(page) |> Gt.Repo.all,
            currentPage: page,
            totalPages: Float.ceil((Project.count(projects) |> Gt.Repo.one) / @page_size),
            lastUpdated: GtDate.timestamp(GtDate.now),
            search: search
        }
    end

    def load_project(id) do
        %{project: Gt.Repo.get!(Project, id)}
    end
end
